
<b>¡Hola Veronica!</b>

Mi nombre es Alejandro Abia y tengo el gusto de revisar tu proyecto.

A continuación, encontrarás mis comentarios en celdas pintadas de tres colores (verde, amarillo y rojo), a manera de semáforo. Por favor, <b>no las borres ni muevas de posición</b> mientras dure el proceso de revisión.

<div class="alert alert-block alert-success">
<b>Éxito</b> <a class="tocSkip"></a>
En celdas verdes encontrarás comentarios en relación a tus aciertos y fortalezas.
</div>
<div class="alert alert-block alert-warning">
<b>Atención</b> <a class="tocSkip"></a>
Utilizaré el color amarillo para llamar tu atención, expresar algo importante o compartirte alguna idea de valor.
</div>
<div class="alert alert-block alert-danger">
<b>A resolver</b> <a class="tocSkip"></a>
En rojo emitiré aquellos puntos que podrían impedir que el proyecto se ejecute correctamente. No son errores, sino oportunidades importantes de mejora.
</div>
<div class="alert alert-block alert-info">
<b>Comentario estudiante</b> <a class="tocSkip"></a>
Si durante la revisión deseas dejarme algún comentario, por favor utiliza celdas azules como esta.
</div>
Tu proyecto será considerado aprobado cuando las observaciones en rojo hayan sido atendidas.  
¡Empecemos!


In [1]:
#Cargar Datos: Importamos las librerías necesarias y leemos los tres archivos.

import pandas as pd
import numpy as np

geo_0 = pd.read_csv('/datasets/geo_data_0.csv')
geo_1 = pd.read_csv('/datasets/geo_data_1.csv')
geo_2 = pd.read_csv('/datasets/geo_data_2.csv')


In [2]:
#iniciamos con revisar la estructura, tipos de datos y posibles valores faltantes.

geo_0.info()
geo_1.info()
geo_2.info()

geo_0.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null 

,id,f0,f1,f2,product
0,txEyH,0.705745,-0.497823,1.221170,105.280062
1,2acmU,1.334711,-0.340164,4.365080,73.037750
2,409Wp,1.022732,0.151990,1.419926,85.265647
3,iJLyR,-0.032172,0.139033,2.978566,168.620776
4,Xdl7t,1.988431,0.155413,4.751769,154.036647


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [2] hiciste un chequeo inicial muy saludable: <code>info()</code> para tipos y conteos no nulos, y <code>head()</code> para ver ejemplos. Esta combinación es una buena práctica porque previene errores “silenciosos” (por ejemplo, columnas numéricas leídas como texto). Mantener esta rutina al inicio de cada proyecto suele ahorrar mucho tiempo después.
</div>

In [3]:
# buscamos valores nulos

geo_0.isna().sum()
geo_1.isna().sum()
geo_2.isna().sum()


id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [3] verificaste valores nulos con <code>isna().sum()</code>. Confirmar que no hay faltantes es importante porque muchos modelos (incluida regresión lineal) no aceptan NaN directamente y, además, los NaN pueden sesgar métricas si se imputan sin cuidado. Este paso deja tu pipeline más confiable.
</div>

In [4]:
# también buscamos valores duplicados

geo_0.duplicated().sum()
geo_1.duplicated().sum()
geo_2.duplicated().sum()


0

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [4] revisaste duplicados. Aunque aquí no aparezcan, es un control valioso porque duplicados pueden inflar artificialmente el desempeño del modelo (se “parece” demasiado a lo que ya vio) y también distorsionar el cálculo de beneficios al contar pozos repetidos. Buen hábito.
</div>

resumen: Cada dataset debe contener:

1.- id → identificador del pozo

2.- f0, f1, f2 → variables predictoras

3.- product → volumen de reservas (target)

Verificamos que:

--No haya valores nulos.

--No haya tipos incorrectos.

--No haya duplicados problemáticos.

In [5]:
#como el id no aporta valor predictivo, lo eliminamos antes de empezar a trabajar en el modelado

geo_0 = geo_0.drop('id', axis=1)
geo_1 = geo_1.drop('id', axis=1)
geo_2 = geo_2.drop('id', axis=1)


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [5] eliminaste <code>id</code> antes del modelado, y tu razonamiento es correcto: es un identificador, no una característica con señal predictiva real. Quitar variables “solo identificadoras” evita que el modelo aprenda relaciones espurias (como memorizar) y mantiene el enfoque en <code>f0</code>, <code>f1</code> y <code>f2</code>.
</div>

Con esto los datasets están listos para dividirse en entrenamiento y validación.

In [6]:
from sklearn.model_selection import train_test_split

# Definir features y target
X_0 = geo_0.drop('product', axis=1)
y_0 = geo_0['product']

X_train_0, X_valid_0, y_train_0, y_valid_0 = train_test_split(
    X_0,
    y_0,
    test_size=0.25,
    random_state=42
)

print(X_train_0.shape)
print(X_valid_0.shape)
print(y_train_0.shape)
print(y_valid_0.shape)


(75000, 3)
(25000, 3)
(75000,)
(25000,)


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [6] hiciste la separación feature/target y el split train/valid con <code>random_state</code>. Esto es clave porque hace tus resultados reproducibles (otra persona corre el notebook y ve lo mismo) y porque evaluar en validación ayuda a medir desempeño “fuera de muestra”, que es justo lo que se necesita antes de hablar de ganancias esperadas.
</div>

Separamos las variables predictoras y la variable objetivo
Usar random_state para que el resultado sea reproducible
Verificación rápida de tamaños
Se observan:
75% de los datos en entrenamiento
25% en validación
Se fijó random_state=42 para garantizar reproducibilidad de los resultados.

In [7]:
from sklearn.linear_model import LinearRegression

model_0 = LinearRegression()

model_0.fit(X_train_0, y_train_0)


LinearRegression()

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [7] elegiste regresión lineal, que en este proyecto es una elección muy adecuada porque el objetivo es replicar un baseline estable y fácil de interpretar. Además, al no requerir escalado obligatorio de variables, reduce fricción y permite enfocarte en la parte de negocio (selección de pozos y riesgo).
</div>

In [8]:
predictions_0 = model_0.predict(X_valid_0)


In [9]:
results_0 = pd.DataFrame({
    'actual': y_valid_0.reset_index(drop=True),
    'predicted': predictions_0
})

results_0.head()


,actual,predicted
0,122.073350,101.901017
1,48.738540,78.217774
2,131.338088,115.266901
3,88.327757,105.618618
4,36.959266,97.980185


Se entrenó un modelo de regresión lineal utilizando las variables f0, f1 y f2 como predictores del volumen de reservas (product).

Posteriormente, se generaron predicciones sobre el conjunto de validación para evaluar el desempeño del modelo en datos no utilizados durante el entrenamiento.

In [10]:
y_valid_0 = y_valid_0.reset_index(drop=True)
predictions_0 = pd.Series(predictions_0)

results_0 = pd.DataFrame({
    'actual': y_valid_0,
    'predicted': predictions_0
})
results_0.head()


,actual,predicted
0,122.073350,101.901017
1,48.738540,78.217774
2,131.338088,115.266901
3,88.327757,105.618618
4,36.959266,97.980185


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [10] alineaste índices antes de construir el DataFrame de resultados. Esta atención al detalle es importante porque un desalineamiento de índices puede llevar a comparar predicciones con “reales” equivocados, y eso arruina tanto RMSE como el cálculo de beneficios. Muy bien asegurada la consistencia.
</div>

Se almacenaron las predicciones del modelo y los valores reales del conjunto de validación en una estructura tabular alineada por índice.

Esto garantiza consistencia en los siguientes cálculos de:

selección de pozos con mayor predicción,

estimación de beneficios,

simulaciones de riesgo mediante bootstrapping.

In [11]:
mean_predicted_0 = results_0['predicted'].mean()
print("Volumen medio predicho:", mean_predicted_0)


Volumen medio predicho: 92.39879990657768


In [12]:
from sklearn.metrics import mean_squared_error
import numpy as np

rmse_0 = np.sqrt(mean_squared_error(results_0['actual'], results_0['predicted']))
print("RMSE:", rmse_0)


RMSE: 37.75660035026169


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [12] calculaste RMSE, que es una métrica natural para regresión cuando quieres penalizar más los errores grandes. Y lo conectaste bien con el impacto económico: en este tipo de decisión, un error de predicción no es solo “técnico”, se traduce en elegir pozos menos rentables. Buena lectura del contexto.
</div>

Se calculó el promedio de reservas estimadas para evaluar el nivel general de producción esperado en la región.
Posteriormente, se calculó el RMSE para medir la precisión del modelo sobre datos no vistos.
Un RMSE bajo indica que el modelo predice con mayor precisión el volumen de reservas. Sin embargo, el análisis final deberá considerar también el impacto económico de dichas predicciones.

Conclusión provisional
--El modelo genera predicciones razonables.
--Existe error inherente que puede impactar directamente el beneficio.
--La decisión final no se debe basar solo en el promedio predicho, sino en la distribución de ganancias y riesgo (bootstrapping).

Todavía no podemos decidir región.

Primero debemos:
--Repetir el proceso para geo_data_1 y geo_data_2.
--Comparar beneficios esperados.
--Medir riesgo de pérdida < 2.5%.

In [13]:
#entrenamiento y evaluación para la region 1 y 2
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd

def train_and_evaluate(data):

    # Separar variables
    X = data.drop('product', axis=1)
    y = data['product']
    
    X_train, X_valid, y_train, y_valid = train_test_split(
        X, y, test_size=0.25, random_state=42
    )
    
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    predictions = model.predict(X_valid)
    
    mean_predicted = predictions.mean()
    rmse = np.sqrt(mean_squared_error(y_valid, predictions))
    
    results = pd.DataFrame({
        'actual': y_valid.reset_index(drop=True),
        'predicted': predictions
    })
    
    return mean_predicted, rmse, results


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [13] encapsulaste el entrenamiento y la evaluación en una función. Esto mejora mantenibilidad y reduce duplicación, algo especialmente útil cuando comparas regiones con el mismo proceso. Como continuidad, podrías devolver también el modelo entrenado si más adelante quisieras inspeccionar coeficientes o reutilizarlo, pero tal como está, cumple muy bien el objetivo.
</div>

In [14]:

mean_0, rmse_0, results_0 = train_and_evaluate(geo_0)
mean_1, rmse_1, results_1 = train_and_evaluate(geo_1)
mean_2, rmse_2, results_2 = train_and_evaluate(geo_2)



In [15]:
print("=== REGIÓN 0 ===")
print(f"Volumen medio predicho: {mean_0:.2f}")
print(f"RMSE: {rmse_0:.2f}")

print("\n=== REGIÓN 1 ===")
print(f"Volumen medio predicho: {mean_1:.2f}")
print(f"RMSE: {rmse_1:.2f}")

print("\n=== REGIÓN 2 ===")
print(f"Volumen medio predicho: {mean_2:.2f}")
print(f"RMSE: {rmse_2:.2f}")

=== REGIÓN 0 ===
Volumen medio predicho: 92.40
RMSE: 37.76

=== REGIÓN 1 ===
Volumen medio predicho: 68.71
RMSE: 0.89

=== REGIÓN 2 ===
Volumen medio predicho: 94.77
RMSE: 40.15


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [15] presentaste una comparación clara entre regiones (media predicha y RMSE). Se nota especialmente el caso de la Región 1 con RMSE muy bajo, lo que ya anticipa (sin decidir todavía) que el modelo ahí es más confiable. Este tipo de lectura comparativa es exactamente lo que se busca antes de pasar al análisis financiero.
</div>

In [16]:
# Parámetros del negocio
BUDGET = 100_000_000          # 100 millones USD
WELLS_TO_DEVELOP = 200        # pozos a desarrollar
SAMPLE_SIZE = 500             # puntos explorados por región
TOP_WELLS = 200               # mejores pozos seleccionados para invertir

REVENUE_PER_BARREL = 4.5      # USD por barril
REVENUE_PER_UNIT = 4500       # USD por 1 unidad (miles de barriles)

# Inversión promedio por pozo
BUDGET_PER_WELL = BUDGET / WELLS_TO_DEVELOP  # 500,000 USD

# Umbral mínimo de producción para no perder (en unidades)
# Cada unidad genera 4500 USD, entonces:
MIN_UNITS_FOR_BREAKEVEN = BUDGET_PER_WELL / REVENUE_PER_UNIT

BUDGET_PER_WELL, MIN_UNITS_FOR_BREAKEVEN


(500000.0, 111.11111111111111)

<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [16] definiste parámetros de negocio con variables (presupuesto, pozos, ingresos por unidad) y calculaste el umbral de equilibrio por pozo. Eso conecta datos y decisión de forma muy pedagógica: traduces el problema a “cuántas unidades necesito para no perder”. Esta trazabilidad hace que tus conclusiones sean defendibles.
</div>

In [17]:
BOOTSTRAP_SAMPLES = 1000
CONFIDENCE_LEVEL = 0.95
MAX_LOSS_RISK = 0.025   # 2.5%


Se definieron variables separadas para presupuesto, número de pozos, tamaño de muestra, ingresos por unidad y parámetros de simulación (bootstrapping). Esto evita números mágicos en el código, mejora legibilidad y asegura trazabilidad con las condiciones del negocio.

In [18]:
print("Promedio Región 0:", geo_0['product'].mean())
print("Promedio Región 1:", geo_1['product'].mean())
print("Promedio Región 2:", geo_2['product'].mean())


Promedio Región 0: 92.50000000000001
Promedio Región 1: 68.82500000000002
Promedio Región 2: 95.00000000000004


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [18] comparaste promedios reales por región y luego interpretaste correctamente por qué el promedio no basta para decidir. Esta es una idea central del proyecto: como vas a elegir los mejores 200, lo relevante es la cola alta de la distribución (los mejores pozos), no el centro. Muy buena intuición estadística aplicada al negocio.
</div>

Tecnicamente, ninguna región alcanza el punto de equilibrio en promedio.
Esto significa:
Si se desarrollaran pozos de manera aleatoria, el proyecto perdería dinero en las tres regiones.
La Región 1 es claramente la más débil estructuralmente (muy por debajo del umbral).
Las regiones 0 y 2 están más cerca del punto de equilibrio, pero siguen por debajo.

El análisis demuestra que la rentabilidad del proyecto no puede basarse en promedios regionales, ya que ninguna región supera el nivel mínimo necesario para evitar pérdidas.

Por lo tanto:
--La viabilidad depende completamente de seleccionar los 200 pozos con mayor volumen estimado.
--La calidad del modelo predictivo es determinante.
--El análisis de riesgo mediante bootstrapping será clave para tomar la decisión final.

Conclusión: Preparación para el cálculo del beneficio

El análisis previo muestra que ninguna región supera el umbral mínimo de producción requerido (111.1 mil barriles por pozo) en promedio. Por lo tanto, la rentabilidad del proyecto no puede basarse en valores promedio regionales, sino en la correcta selección de los pozos con mayor potencial productivo.

Para calcular el beneficio correctamente, es necesario:
1.-Utilizar las predicciones generadas por el modelo de regresión lineal.
2.-Seleccionar los 200 pozos con mayor volumen estimado en cada región.
3.-Calcular el volumen total esperado sumando las reservas reales correspondientes a esos pozos seleccionados.
4.-Multiplicar ese volumen por el ingreso por unidad (4500 USD).
5.-Restar el presupuesto total invertido (100 millones USD).

Esto permite estimar el beneficio potencial bajo una estrategia de selección óptima, alineada con las condiciones del negocio.

Además, dado que el modelo tiene error (RMSE distinto de cero), el beneficio calculado representa una estimación y deberá complementarse con un análisis de riesgo mediante bootstrapping para evaluar la probabilidad de pérdidas.

La lógica financiera es:
𝐺anancia=(Volumen total x 45000)-100,000,000
Importante:
Seleccionamos por predicción, pero calculamos con el valor real (actual).

In [19]:
def calculate_profit(results, top_n, revenue_per_unit, budget):
    
    # Ordenar por predicción descendente
    top_wells = results.sort_values(by='predicted', ascending=False).head(top_n)
    
    # Sumar volumen real de esos pozos
    total_volume = top_wells['actual'].sum()
    
    # Calcular ingreso
    revenue = total_volume * revenue_per_unit
    
    # Ganancia
    profit = revenue - budget
    
    return profit, total_volume


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [19] tu función de ganancia respeta el principio correcto del proyecto: seleccionar por <code>predicted</code> pero calcular ingreso con <code>actual</code>. Eso evita “hacer trampa” usando el target para seleccionar y refleja el mundo real: decides con pronósticos, pero el dinero se gana/pierde con lo que realmente hay.
</div>

In [20]:
profit_0, volume_0 = calculate_profit(results_0, 200, REVENUE_PER_UNIT, BUDGET)
profit_1, volume_1 = calculate_profit(results_1, 200, REVENUE_PER_UNIT, BUDGET)
profit_2, volume_2 = calculate_profit(results_2, 200, REVENUE_PER_UNIT, BUDGET)


In [21]:
print("Región 0")
print("Volumen total:", volume_0)
print("Ganancia:", profit_0)

print("\nRegión 1")
print("Volumen total:", volume_1)
print("Ganancia:", profit_1)

print("\nRegión 2")
print("Volumen total:", volume_2)
print("Ganancia:", profit_2)


Región 0
Volumen total: 29686.9802543604
Ganancia: 33591411.14462179

Región 1
Volumen total: 27589.081548181137
Ganancia: 24150866.966815114

Región 2
Volumen total: 27996.82613194247
Ganancia: 25985717.59374112


<div class="alert alert-block alert-warning">
<b>Oportunidad de mejora</b> <a class="tocSkip"></a><br>
En tu explicación financiera antes de la celda [19] aparece la fórmula con <code>45000</code>, mientras que en el código usas <code>REVENUE_PER_UNIT = 4500</code>. Es un detalle crítico porque un cero extra cambia la ganancia por un factor de 10 y podría llevar a una interpretación equivocada, aunque tus cálculos en código sí parecen consistentes con 4500. Conviene unificarlo (texto y variables) para que la narrativa y los resultados cuenten la misma historia.
</div>

Región 0 lidera claramente
Mayor volumen total.
Mayor ganancia estimada.
Diferencia de casi 8–9 millones frente a las otras regiones.
Como resultado previo, la Región 0 es la más atractiva.

Región 1 es la más débil
Ya era la más baja en promedio general.
También queda última en ganancia.
Consistencia negativa estructural.

Región 2 es intermedia
Mejora respecto a la 1.
Pero no alcanza a la 0.

In [ ]:
Si tomáramos la decisión solo con el beneficio estimado de los 200 mejores pozos, la Región 0 sería la elección óptima.

Pero aquí viene el punto clave del proyecto:
Esto asume que las predicciones son perfectamente confiables.

Sabemos que el modelo tiene RMSE.
Eso implica incertidumbre.

Y el negocio exige:
Riesgo de pérdida < 2.5%

Por lo tanto, aún no podemos decidir oficialmente.

Ahora necesitamos aplicar bootstrapping para medir:
Beneficio promedio
Intervalo de confianza 95%
Probabilidad de pérdida

In [22]:
def select_top_wells(results, top_n):
    top_wells = results.sort_values(by='predicted', ascending=False).head(top_n)
    return top_wells

top_0 = select_top_wells(results_0, 200)
top_1 = select_top_wells(results_1, 200)
top_2 = select_top_wells(results_2, 200)


In [23]:
volume_0 = top_0['actual'].sum()
volume_1 = top_1['actual'].sum()
volume_2 = top_2['actual'].sum()

print("Volumen Región 0:", volume_0)
print("Volumen Región 1:", volume_1)
print("Volumen Región 2:", volume_2)


Volumen Región 0: 29686.9802543604
Volumen Región 1: 27589.081548181137
Volumen Región 2: 27996.82613194247


In [24]:
#Las predicciones de los 200 pozos están en:
top_0['predicted']
top_1['predicted']
top_2['predicted']


21852    170.529209
10722    169.673332
6209     165.300724
8203     164.613896
8042     163.964000
            ...    
11300    142.609491
22066    142.562875
1232     142.480528
18966    142.373217
13343    142.302492
Name: predicted, Length: 200, dtype: float64

In [26]:
#Y los valores reales en:
top_0['actual']
top_1['actual']
top_2['actual']


21852    101.225039
10722    151.655778
6209      92.947333
8203      97.775979
8042     122.460897
            ...    
11300    179.654566
22066    142.101687
1232     121.983862
18966    184.895101
13343    151.695643
Name: actual, Length: 200, dtype: float64

Los 200 pozos fueron seleccionados con base en las predicciones del modelo. Sin embargo, al comparar con los valores reales se observa variabilidad significativa entre lo estimado y lo observado, lo cual refleja el error inherente del modelo (RMSE).

Esta diferencia confirma la necesidad de realizar un análisis de riesgo mediante simulación bootstrap para evaluar la probabilidad de pérdidas antes de tomar una decisión final de inversión.

Análisis comparativo
Región 0
--Mayor volumen total seleccionado.
--Mayor ganancia estimada.
--Ventaja de aproximadamente 7–9 millones frente a las otras regiones.
--Estructuralmente más fuerte.

Región 1
La más débil desde el promedio inicial.
--Menor volumen total.
--Menor rentabilidad estimada.

Región 2
--Intermedia.
--Mejor que la 1, pero claramente inferior a la 0.

Conclusión preliminar (sin considerar riesgo)
Si la decisión se basara únicamente en la ganancia potencial determinada, la Región 0 sería la mejor opción para desarrollar los 200 pozos petrolíferos.

Porque la región presenta, como ya dijimos:
--Mayor volumen total real entre los pozos seleccionados.
--Mayor beneficio esperado.
--Mayor distancia frente al punto de equilibrio.

Pero, sabemos que existe error (RMSE), y el negocio exige: Riesgo de pérdida inferior al 2.5%
Por lo tanto, aunque Región 0 lidera en ganancia estimada, la decisión final dependerá del análisis de riesgo mediante bootstrapping.

In [27]:
def bootstrap_profit(results, n_samples, sample_size, top_n, revenue_per_unit, budget):
    
    profits = []
    
    for i in range(n_samples):
        
        # Muestra aleatoria con reemplazo
        sample = results.sample(n=sample_size, replace=True, random_state=None)
        
        # Seleccionar los mejores 200 por predicción
        top_wells = sample.sort_values(by='predicted', ascending=False).head(top_n)
        
        # Calcular volumen real
        total_volume = top_wells['actual'].sum()
        
        # Calcular ganancia
        revenue = total_volume * revenue_per_unit
        profit = revenue - budget
        
        profits.append(profit)
    
    return pd.Series(profits)


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [27] implementaste bootstrapping de manera coherente con el planteamiento: muestreo con reemplazo, selección del top por predicción y cálculo de ganancia con valores reales. Esto es valioso porque en vez de quedarte con un único número de beneficio, obtienes una distribución, que es la base para hablar de riesgo e intervalos de confianza.
</div>

In [28]:
profits_0 = bootstrap_profit(results_0, 1000, 500, 200, REVENUE_PER_UNIT, BUDGET)
profits_1 = bootstrap_profit(results_1, 1000, 500, 200, REVENUE_PER_UNIT, BUDGET)
profits_2 = bootstrap_profit(results_2, 1000, 500, 200, REVENUE_PER_UNIT, BUDGET)


In [29]:
#Para cada región necesitamos:
#Beneficio promedio
#Intervalo de confianza 95%
#Riesgo de pérdida (probabilidad de ganancia negativa)

def analyze_bootstrap(profits):
    
    mean_profit = profits.mean()
    
    lower = profits.quantile(0.025)
    upper = profits.quantile(0.975)
    
    risk = (profits < 0).mean()
    
    return mean_profit, lower, upper, risk


In [30]:
mean_0, lower_0, upper_0, risk_0 = analyze_bootstrap(profits_0)
mean_1, lower_1, upper_1, risk_1 = analyze_bootstrap(profits_1)
mean_2, lower_2, upper_2, risk_2 = analyze_bootstrap(profits_2)

print("Región 0:", mean_0, lower_0, upper_0, risk_0)
print("Región 1:", mean_1, lower_1, upper_1, risk_1)
print("Región 2:", mean_2, lower_2, upper_2, risk_2)


Región 0: 4016996.416045138 -1772579.4836079369 9174619.159286814 0.07
Región 1: 4337088.347034351 617217.9955411379 8530622.867929209 0.011
Región 2: 3841283.337274503 -1201121.3669819671 8953939.889014395 0.069


<div class="alert alert-block alert-success">
<b>Acierto o fortaleza</b> <a class="tocSkip"></a><br>
En la celda [30] cerraste el análisis con los tres elementos que pide el negocio: beneficio esperado, intervalo al 95% y probabilidad de pérdida. Y tu decisión final está bien alineada con la restricción (< 2.5%): aunque la Región 0 gane más en el cálculo “determinista”, la Región 1 es la que cumple política de riesgo. Esa es una conclusión madura: prioriza reglas del negocio sobre el número más alto.
</div>

Resumen de los resultados obtenidos:
Región 0
Aunque tenía la mayor ganancia (33 M),
al incorporar incertidumbre:
Riesgo = 7%. No procede (supera 2.5%)
Intervalo incluye pérdidas. No procede
Queda descartada por política de riesgo.

Región 2
Situación similar:
Riesgo ≈ 6.9%. No procede
Intervalo incluye pérdidas. No procede
También descartada.

Región 1
Aquí está la diferencia clave:
Riesgo = 1.1%. Si procede (menor a 2.5%)
Intervalo de confianza completamente positivo. Si procede
Mayor beneficio promedio bajo incertidumbre. Si procede
Aunque en el análisis previo parecía menos atractiva,
es la única región que cumple la restricción de riesgo del negocio.

Decisión Final
La región recomendada para el desarrollo de los 200 pozos es:
Región 1
Justificación:
--Cumple el criterio de riesgo (< 2.5%)
--Presenta el mayor beneficio promedio bajo simulación
--Es la única región con intervalo de confianza completamente positivo y,
--Ofrece mejor balance rentabilidad-riesgo

Conclusión final
Tras aplicar la técnica de bootstrapping con 1000 muestras para estimar la distribución de beneficios en cada región, se obtuvieron los siguientes resultados:
Las regiones 0 y 2 presentan probabilidades de pérdida del 7% y 6.9% respectivamente, superando el umbral máximo permitido del 2.5%. Además, sus intervalos de confianza incluyen valores negativos, lo que indica riesgo significativo de pérdidas.
Por el contrario, la Región 1 presenta una probabilidad de pérdida del 1.1%, inferior al límite establecido. Su intervalo de confianza al 95% es completamente positivo y muestra el mayor beneficio promedio bajo incertidumbre.
Aunque en el análisis determinístico inicial la Región 0 parecía más rentable, al incorporar el riesgo mediante simulación estadística, la Región 1 resulta ser la opción más sólida desde una perspectiva financiera y de gestión del riesgo.
Por lo tanto, se recomienda seleccionar la Región 1 para el desarrollo de los 200 pozos petrolíferos.

¿Coincide mi elección con la anterior? NO

<div class="alert alert-block alert-success">
<b>Comentario final</b> <a class="tocSkip"></a><br>
¡Muy buen trabajo, Veronica! A lo largo del proyecto mostraste fortalezas muy claras:<br><br>
• Realizaste una revisión inicial sólida de estructura, tipos, nulos y duplicados, lo que da confianza en el dataset<br>
• Tomaste una buena decisión al eliminar <code>id</code> para evitar ruido o señales espurias en el modelado<br>
• Separaste correctamente entrenamiento/validación con reproducibilidad, cuidando el método experimental<br>
• Construiste un baseline de regresión lineal bien aplicado y medido con RMSE<br>
• Alineaste índices y resultados de manera cuidadosa, evitando errores sutiles en métricas y ganancias<br>
• Estandarizaste el proceso para las tres regiones con funciones reutilizables y claras<br>
• Tradujiste el problema al lenguaje de negocio (presupuesto, umbral de equilibrio), conectando datos con decisión<br>
• Calculaste beneficio de forma correcta: selección por predicción y evaluación con valores reales<br>
• Implementaste bootstrapping de forma coherente para estimar incertidumbre e intervalo de confianza<br>
• Tomaste una decisión final alineada con la restricción de riesgo, priorizando política del negocio sobre el máximo beneficio puntual<br>
<br>
¡Felicidades!
</div>